### PR TITLE
Remove all remaining uses of open_on_apply

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

Part of the Epic to remove the concept of `open_on_apply` from the application.

The places where `open_on_apply` is used are the following:
 - places where the application did not want to treat a course as open because it was not yet moved from UCAS to Apply
 - places where it was believed that `open_on_apply` was used by providers to disable courses arbitrarily.

## Changes proposed in this pull request

 - Change Course open_on_apply to `open` where appropriate in the `CandidateInterface`
 - Remove `open_on_apply` where it serves no purpose anymore
 - Replace `open_on_apply` with `open` where it makes sense to do so
 - Replace `open_on_apply` with `application_status_open` where it makes sense to do so

## Guidance to review

I have made comments on some changes to explain, clarify or to ask for direction on a change.

## Link to Trello card

[Trello Ticket](https://trello.com/c/Xpo5TFJQ/1434-openonapply-replace-openonapply-with-open-in-candidateinterface)

### CODE REVIEW

File: spec/forms/candidate_interface/pick_provider_form_spec.rb [hunk 1]
```diff
@@ -4,10 +4,10 @@
...
-      create(:course, open_on_apply: true, exposed_in_find: false, provider: create(:provider, name: 'School with disabled courses'))
-      create(:course, open_on_apply: true, exposed_in_find: true, applications_open_from: Time.zone.today, provider: create(:provider, name: 'School with courses'))
-      create(:course, open_on_apply: true, exposed_in_find: true, applications_open_from: Time.zone.tomorrow, provider: create(:provider, name: 'School with courses but not open for applications'))
-      create(:course, open_on_apply: true, exposed_in_find: true, recruitment_cycle_year: RecruitmentCycle.previous_year)
+      create(:course, :open, exposed_in_find: false, provider: create(:provider, name: 'School with disabled courses'))
+      create(:course, :open, exposed_in_find: true, applications_open_from: Time.zone.today, provider: create(:provider, name: 'School with courses'))
+      create(:course, :open, exposed_in_find: true, applications_open_from: Time.zone.tomorrow, provider: create(:provider, name: 'School with courses but not open for applications'))
+      create(:course, :open, exposed_in_find: true, recruitment_cycle_year: RecruitmentCycle.previous_year)
```
Comment (inulty-dfe) [inline spec/forms/candidate_interface/pick_provider_form_spec.rb:10]: These don't actually need to be `open`. Should we change the code to filter open courses and not just exposed in find?

**Review by inulty-dfe — COMMENTED** [review]
Leaving some comments ot help reviewers of the code.

There are some questions too.

File: spec/forms/candidate_interface/pick_provider_form_spec.rb (hunk 1 above)
Comment (inulty-dfe) [reply spec/forms/candidate_interface/pick_provider_form_spec.rb:10]: The pick form filters out courses that aren't `exposed_in_find`. Should it also filter out courses that are closed by providers?

https://github.com/DFE-Digital/apply-for-teacher-training/blob/open_on_apply_candidate_interface/app/forms/candidate_interface/pick_provider_form.rb#L8-L13

File: app/models/course.rb [hunk 2]
```diff
@@ -174,15 +169,6 @@ def ratifying_provider
...
-
-    update!(
-      open_on_apply: true,
-      opened_on_apply_at: Time.zone.now,
-    )
-  end
-
   def ske_graduation_cutoff_date
     start_date - SKE_GRADUATION_CUTOFF_THRESHOLD
   end
```
Comment (tomas-stefano) [inline app/models/course.rb:174]: Maybe is not related to this PR but related to the open feature.

Can we use `open_for_applications?` in the `open?` method?

```ruby
  def open?
    applications_open_from.present? && applications_open_from <= Time.zone.today && exposed_in_find && application_status_open?
  end

  def open_for_applications?
    applications_open_from <= Time.zone.today
  end
```

File: spec/forms/candidate_interface/pick_provider_form_spec.rb (hunk 1 above)
Comment (tomas-stefano) [reply spec/forms/candidate_interface/pick_provider_form_spec.rb:10]: I think it should, although I assume there shouldn't be any courses that are closed and exposed in find? Although I think it should.

Comment (inulty-dfe) [reply spec/forms/candidate_interface/pick_provider_form_spec.rb:10]: The pick provider form should return providers without valid courses. The UI will handle when the candidate chooses and invalid provider and sends them to an error page. (whether that is a good idea or not.)

File: app/models/course.rb (hunk 2 above)
Comment (inulty-dfe) [reply app/models/course.rb:174]: I'll make a note to do this in my next PR which aims to make use of `application_status` in the areas it has not been introduced.

File: app/queries/get_open_courses_by_provider_and_region.rb [hunk 3]
```diff
@@ -3,7 +3,6 @@ class GetOpenCoursesByProviderAndRegion
 
   def self.call
     Course
-      .open_on_apply
```
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/queries/get_open_courses_by_provider_and_region.rb:6]: Done

File: spec/forms/candidate_interface/pick_provider_form_spec.rb (hunk 1 above)
Comment (inulty-dfe) [reply spec/forms/candidate_interface/pick_provider_form_spec.rb:10]: We've had a discussion around the concepts of `open` and `exposed_in_find`. Also around the logic of presenting providers in this dropdown when they are not available for applications. The changes in this PR are ok to go ahead considering these discussions.

**Pull request merged**